### PR TITLE
fix FindOpenEXR to respect OPENEXR_ROOT

### LINKED ISSUE #13403: Provide method to find user-installed OpenEXR
##### System information (version)

- OpenCV => 2.4+ (All versions since 2012)
- Operating System / Platform => Debian 64 Bit
- Compiler => GCC 6

###### Background
I found this issue while trying to use the Conan package manager. I realise that the OpenCV does not provide any official package manager support, but I believe this fix could help others too. https://github.com/conan-community/community/issues/114

###### Description
In the FindOpenEXR.cmake file (https://github.com/opencv/opencv/blob/master/cmake/OpenCVFindOpenEXR.cmake), there is support for using a variable called 'OPENEXR_ROOT' to find a version of OpenEXR installed in a user-defined location. Currently this is a cache variable with a default of C:\Deploy and is only defined/useable in Windows:

```.cmake
if(WIN32)
    SET(OPENEXR_ROOT "C:/Deploy" CACHE STRING "Path to the OpenEXR \"Deploy\" folder")
    if(CMAKE_CL_64)
        SET(OPENEXR_LIBSEARCH_SUFFIXES x64/Release x64 x64/Debug)
    elseif(MSVC)
        SET(OPENEXR_LIBSEARCH_SUFFIXES Win32/Release Win32 Win32/Debug)
    endif()
else()
    set(OPENEXR_ROOT "")
endif()
```

 It would be great if it could also be a cache variable for other platforms, we can keep the default to "", as it is being set to now.

The second thing that would be required is to change the following blob:
```.cmake
SET(LIBRARY_PATHS
    /usr/lib
    /usr/local/lib
    /sw/lib
    /opt/local/lib
    "${ProgramFiles_ENV_PATH}/OpenEXR/lib/static"
    "${OPENEXR_ROOT}/lib")
 ```
to

```.cmake
SET(LIBRARY_PATHS
    "${OPENEXR_ROOT}/lib"
    /usr/lib
    /usr/local/lib
    /sw/lib
    /opt/local/lib
    "${ProgramFiles_ENV_PATH}/OpenEXR/lib/static")
 ```
i.e. if we explicitly define OPENEXR_ROOT then we would expect to search that location first before any system installed versions. This is in particular important for use with package managers (which is where I ran into the problem). 

If there are problems using that OPENEXR_ROOT for whatever reason, maybe it would be OK to use a new variable instead e.g. OPENEXR_USER_ROOT, which could default?

### PR DESCRIPTION
**Pull request by SSE4**
resolves #13403

### This pullrequest changes

- if CMake variable `OPENEXR_ROOT` is set, it has the top priority, so user-supplied OpenEXR version always wins against system-provided or one provided by OpenCV itself
- do not try to override user-specified `OPENEXR_ROOT` to hard-coded paths (like **C:\deploy**)
- OpenEXR libraries may use various versioning schemes, depending on compiler or configuration (debug vs release, static vs shared, etc.) for instance, `Half` library may have the following names in OpenEXR 2.3: `Half-2_3 Half-2_3_s Half-2_3_d Half-2_3_s_d Half Half_s Half_d Half_s_d`
- parse `OpenEXRConfig.h` to find OpenEXR version number in order to know versioned library names

changes above basically allow to find OpenEXR supplied by various package managers (e.g. conan, vcpkg, hunter, cget, cppan, brew, etc.)

### CODE REVIEW

File: cmake/OpenCVFindOpenEXR.cmake [hunk 1]
```diff
@@ -13,57 +13,77 @@ SET(OPENEXR_LIBRARIES "")
...
+        if(CMAKE_CL_64)
+            SET(OPENEXR_LIBSEARCH_SUFFIXES x64/Release x64 x64/Debug)
+        elseif(MSVC)
+            SET(OPENEXR_LIBSEARCH_SUFFIXES Win32/Release Win32 Win32/Debug)
+        endif()
+    else()
+        set(OPENEXR_ROOT "")
     endif()
-else()
-    set(OPENEXR_ROOT "")
```
Comment (alalek) [inline cmake/OpenCVFindOpenEXR.cmake]: `SET(OPENEXR_ROOT ... CACHE ...)` is enough. Cache values specified by users are not overwritten by these statements.

Just remove "else()" branch.

Comment (SSE4) [reply cmake/OpenCVFindOpenEXR.cmake]: removed else branch

**Comment by SSE4** [comment]
@alalek I've changed the logic according to your suggestions

**Comment by asmorkalov** [comment]
@SSE4 I tried the solution for Ubuntu 16.04 and it does not work for me. Please correct me, if I do something wrong:
```
$  git clone https://github.com/openexr/openexr.git
$ mkdir openexr_build
$ cd mkdir openexr_build
$ cmake -DCMAKE_INSTALL_PREFIX=/home/alexander/tmp
$ make
$ make install
$ cd ..
$ mkdir opencv_build
$ cmake -DOPENEXR_ROOT=/home/alexander/tmp ../opencv
```
Console log for OpenCV:
```
--   Media I/O: 
--     ZLib:                        /usr/lib/x86_64-linux-gnu/libz.so (ver 1.2.8)
--     JPEG:                        /usr/lib/x86_64-linux-gnu/libjpeg.so (ver 80)
--     WEBP:                        build (ver encoder: 0x020e)
--     PNG:                         /usr/lib/x86_64-linux-gnu/libpng.so (ver 1.2.54)
--     TIFF:                        /usr/lib/x86_64-linux-gnu/libtiff.so (ver 42 / 4.0.6)
--     JPEG 2000:                   /usr/lib/x86_64-linux-gnu/libjasper.so (ver 1.900.1)
--     OpenEXR:                     build (ver 2.3.0)
--     HDR:                         YES
--     SUNRASTER:                   YES
--     PXM:                         YES
--     PFM:                         YES
```

CMakeCache.txt content:
```
alexander@asmorkal-desktop:~/Projects/opencv-build-pr$ cat ./CMakeCache.txt | grep -in openexr
125://Build openexr from source
126:BUILD_OPENEXR:BOOL=OFF
780:OPENEXR_HALF_LIBRARY:FILEPATH=OPENEXR_HALF_LIBRARY-NOTFOUND
783:OPENEXR_IEX_LIBRARY:FILEPATH=OPENEXR_IEX_LIBRARY-NOTFOUND
786:OPENEXR_ILMTHREAD_LIBRARY:FILEPATH=OPENEXR_ILMTHREAD_LIBRARY-NOTFOUND
789:OPENEXR_IMATH_LIBRARY:FILEPATH=OPENEXR_IMATH_LIBRARY-NOTFOUND
792:OPENEXR_INCLUDE_PATH:PATH=/usr/include/OpenEXR
795:OPENEXR_ROOT:UNINITIALIZED=/home/alexander/tmp
986://Include ILM support via OpenEXR
987:WITH_OPENEXR:BOOL=ON
1320:openexr_BINARY_DIR:STATIC=/home/alexander/Projects/opencv-build-pr/3rdparty/openexr
1323:openexr_SOURCE_DIR:STATIC=/home/alexander/Projects/opencv/3rdparty/openexr
3047://ADVANCED property for variable: OPENEXR_HALF_LIBRARY
3048:OPENEXR_HALF_LIBRARY-ADVANCED:INTERNAL=1
3049://ADVANCED property for variable: OPENEXR_IEX_LIBRARY
3050:OPENEXR_IEX_LIBRARY-ADVANCED:INTERNAL=1
3051://ADVANCED property for variable: OPENEXR_ILMTHREAD_LIBRARY
3052:OPENEXR_ILMTHREAD_LIBRARY-ADVANCED:INTERNAL=1
3053://ADVANCED property for variable: OPENEXR_IMATH_LIBRARY
3054:OPENEXR_IMATH_LIBRARY-ADVANCED:INTERNAL=1
```

**Comment by SSE4** [comment]
@asmorkalov steps appear to be correct. I'll take a look and let you know.

**Comment by SSE4** [comment]
@asmorkalov worked for me (`Ubuntu 18.04.2 LTS`):
```
-- Found OpenEXR: /home/sse4/tmp/lib/libIlmImf-2_4.so
...
--   Media I/O:
--     ZLib:                        /usr/lib/x86_64-linux-gnu/libz.so (ver 1.2.11)
--     JPEG:                        libjpeg-turbo (ver 2.0.2-62)
--     WEBP:                        build (ver encoder: 0x020e)
--     PNG:                         /usr/lib/x86_64-linux-gnu/libpng.so (ver 1.6.34)
--     TIFF:                        build (ver 42 - 4.0.10)
--     JPEG 2000:                   build (ver 1.900.1)
--     OpenEXR:                     /home/sse4/tmp/lib/libImath-2_4.so /home/sse4/tmp/lib/libIlmImf-2_4.so /home/sse4/tmp/lib/libIex-2_4.so /home/sse4/tmp/lib/libHalf-2_4.so /home/sse4/tmp/lib/libIlmThread-2_4.so (ver 2_4)
--     HDR:                         YES
--     SUNRASTER:                   YES
--     PXM:                         YES
--     PFM:                         YES
```
```
//Build openexr from source
BUILD_OPENEXR:BOOL=OFF
OPENEXR_HALF_LIBRARY:FILEPATH=/home/sse4/tmp/lib/libHalf-2_4.so
OPENEXR_IEX_LIBRARY:FILEPATH=/home/sse4/tmp/lib/libIex-2_4.so
OPENEXR_ILMIMF_LIBRARY:FILEPATH=/home/sse4/tmp/lib/libIlmImf-2_4.so
OPENEXR_ILMTHREAD_LIBRARY:FILEPATH=/home/sse4/tmp/lib/libIlmThread-2_4.so
OPENEXR_IMATH_LIBRARY:FILEPATH=/home/sse4/tmp/lib/libImath-2_4.so
OPENEXR_INCLUDE_PATH:PATH=/home/sse4/tmp/include/OpenEXR
//The include paths needed to use OpenEXR
OPENEXR_INCLUDE_PATHS:PATH=/home/sse4/tmp/include/OpenEXR
//The libraries needed to use OpenEXR
OPENEXR_LIBRARIES:STRING=/home/sse4/tmp/lib/libImath-2_4.so;/home/sse4/tmp/lib/libIlmImf-2_4.so;/home/sse4/tmp/lib/libIex-2_4.so;/home/sse4/tmp/lib/libHalf-2_4.so;/home/sse4/tmp/lib/libIlmThread-2_4.so
OPENEXR_ROOT:UNINITIALIZED=/home/sse4/tmp
//Include ILM support via OpenEXR
WITH_OPENEXR:BOOL=ON
//ADVANCED property for variable: OPENEXR_HALF_LIBRARY
OPENEXR_HALF_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: OPENEXR_IEX_LIBRARY
OPENEXR_IEX_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: OPENEXR_ILMIMF_LIBRARY
OPENEXR_ILMIMF_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: OPENEXR_ILMTHREAD_LIBRARY
OPENEXR_ILMTHREAD_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: OPENEXR_IMATH_LIBRARY
OPENEXR_IMATH_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: OPENEXR_INCLUDE_PATHS
OPENEXR_INCLUDE_PATHS-ADVANCED:INTERNAL=1
//ADVANCED property for variable: OPENEXR_LIBRARIES
OPENEXR_LIBRARIES-ADVANCED:INTERNAL=1
```

**Comment by asmorkalov** [comment]
@SSE4 I apologize for the delay. I tried the patch with Ubuntu 16.04 and 18.04. It works properly on 18.04, but not on 16.04. Most probably it's issue of CMake 3.5.1. I'll dig a little bit in this today and let you know.

**Comment by SSE4** [comment]
@asmorkalov no problem. I'll try on 16.04 as well (and older CMake) and let you know about my findings, if it's related to Ubuntu version or CMake version.

**Comment by asmorkalov** [comment]
@SSE4, do you have any progress on this?

**Comment by SSE4** [comment]
@asmorkalov on Ubuntu 16.04 I get:
```
CMake Error at CMakeLists.txt:12 (cmake_minimum_required):
  CMake 3.10 or higher is required.  You are running version 3.5.1


-- Configuring incomplete, errors occurred!
```
while building OpenEXR
checking with newer CMake

**Comment by SSE4** [comment]
@asmorkalov results:
works on both Ubuntu 16.04 and Ubuntu 18.04
works with both CMake 3.15.5 and CMake 3.10.2
OpenEXR may only be built with CMake 3.10+ https://github.com/openexr/openexr/commit/032304a556c5ab00748cb55712c84fe131304264

**Comment by asmorkalov** [comment]
@SSE4 I apologize that I misdirect you with CMake versions. I have several versions on my host and I used 3.10+ for both OpenEXR and OpenCV as the first one cannot be built with old CMake. Let me describe what happened on my host and how to fix it.

`FIND_PATH` like other `FIND_` functions set cache variable with path and does not search again, if something was found for the first time. I have some incomplete OpenEXR installation on my host or something that was not detected by your script in system directories. `FIND_PATH` finds headers in system path before hand and always returns the same value ignoring `PATH` option.

I see two things to fix:
- Add `NO_DEFAULT_PATH` option for `FIND_PATH` to search for headers in locations that are defined in function parameters only.
- `UNSET` `OPENEXR_INCLUDE_PATH` and all `_LIBRARY` variables from cache in loop body. Otherwise header files or some library found once earlier is mixed with things found on the next iteration. It's the issue @alalek mentioned before.

**Comment by SSE4** [comment]
@asmorkalov okay, I'll do that adjustments. meanwhile, do you have a command line to reproduce and test the explained case (to get some incomplete OpenEXR installation)?

**Comment by asmorkalov** [comment]
@SSE4 I just tried to install everything related to OpenEXR on my Ubuntu 16.04. I'm not sure if it's complete installation or not, but it was not properly recognized by your code -- only headers were found.

**Comment by SSE4** [comment]
@asmorkalov I've made suggested adjustments

File: cmake/OpenCVFindOpenEXR.cmake [hunk 2]
```diff
@@ -13,62 +13,98 @@ SET(OPENEXR_LIBRARIES "")
 SET(OPENEXR_LIBSEARCH_SUFFIXES "")
 file(TO_CMAKE_PATH "$ENV{ProgramFiles}" ProgramFiles_ENV_PATH)
 
-if(WIN32)
-    SET(OPENEXR_ROOT "C:/Deploy" CACHE STRING "Path to the OpenEXR \"Deploy\" folder")
-    if(CMAKE_CL_64)
-        SET(OPENEXR_LIBSEARCH_SUFFIXES x64/Release x64 x64/Debug)
-    elseif(MSVC)
-        SET(OPENEXR_LIBSEARCH_SUFFIXES Win32/Release Win32 Win32/Debug)
+if(NOT OPENEXR_ROOT)
```
Comment (alalek) [inline cmake/OpenCVFindOpenEXR.cmake]: This condition is not needed - `set(... CACHE)` doesn't override variables be default.

File: cmake/OpenCVFindOpenEXR.cmake [hunk 3]
```diff
@@ -13,62 +13,86 @@ SET(OPENEXR_LIBRARIES "")
...
+        PATH_SUFFIXES ${OPENEXR_LIBSEARCH_SUFFIXES}
+        PATHS "${SEARCH_PATH}/lib" "${SEARCH_PATH}/lib/static")
+ENDMACRO()
 
-FIND_LIBRARY(OPENEXR_HALF_LIBRARY
-    NAMES Half
-    PATH_SUFFIXES ${OPENEXR_LIBSEARCH_SUFFIXES}
-    PATHS ${LIBRARY_PATHS})
+FOREACH(SEARCH_PATH ${SEARCH_PATHS})
+    FIND_PATH(OPENEXR_INCLUDE_PATH ImfRgbaFile.h
```
[the remark this answers is not in the document]
Comment (SSE4) [reply cmake/OpenCVFindOpenEXR.cmake]: great news, thank you very much

File: cmake/OpenCVFindOpenEXR.cmake (hunk 2 above)
Comment (SSE4) [reply cmake/OpenCVFindOpenEXR.cmake]: condition was removed

**Comment by SSE4** [comment]
@alalek rebased onto 3.4

**Pull request merged**